src/synthcity/plugins/core/models/tabular_ddpm/utils.py

```diff
@@ -0,0 +1,216 @@
+    log_probs = torch.where(
+        x < -0.999,
+        log_cdf_plus,
+        torch.where(
+            x > 0.999, log_one_minus_cdf_min, torch.log(cdf_delta.clamp(min=1e-12))
+        ),
+    )
+    if not (log_probs.shape == x.shape):
+        raise AssertionError
+    return log_probs
+
+
+def sum_except_batch(x: Tensor, num_dims: int = 1) -> Tensor:
+    """
+    Sums all dimensions except the first.
+
+    Args:
+        x: Tensor, shape (batch_size, ...)
+        num_dims: int, number of batch dims (default=1)
+
+    Returns:
+        x_sum: Tensor, shape (batch_size,)
+    """
+    return x.reshape(*x.shape[:num_dims], -1).sum(-1)
+
+
+def mean_flat(tensor: Tensor) -> Tensor:
+    """
+    Take the mean over all non-batch dimensions.
+    """
+    return tensor.mean(dim=list(range(1, len(tensor.shape))))
+
+
+def ohe_to_categories(ohe: Tensor, K: np.ndarray) -> Tensor:
+    K = torch.from_numpy(K)
+    indices = torch.cat([torch.zeros((1,)), K.cumsum(dim=0)], dim=0).int().tolist()
+    res = []
+    for i in range(len(indices) - 1):
+        res.append(ohe[:, indices[i] : indices[i + 1]].argmax(dim=1))
+    return torch.stack(res, dim=1)
+
+
+def log_1_min_a(a: Tensor) -> Tensor:
+    return torch.log(1 - a.exp() + 1e-40)
+
+
+def log_add_exp(a: Tensor, b: Tensor) -> Tensor:
+    maximum = torch.max(a, b)
+    return maximum + torch.log(torch.exp(a - maximum) + torch.exp(b - maximum))
+
+
+def extract(a: Tensor, t: Tensor, x_shape: tuple) -> Tensor:
+    b, *_ = t.shape
+    t = t.to(a.device)
+    out = a.gather(-1, t)
+    while len(out.shape) < len(x_shape):
+        out = out[..., None]
+    return out.expand(x_shape)
+
+
+def log_categorical(log_x_start: Tensor, log_prob: Tensor) -> Tensor:
+    return (log_x_start.exp() * log_prob).sum(dim=1)
+
+
+def index_to_log_onehot(x: Tensor, num_classes: np.ndarray) -> Tensor:
+    onehots = []
+    for i in range(len(num_classes)):
+        onehots.append(F.one_hot(x[:, i], num_classes[i]))
+    x_onehot = torch.cat(onehots, dim=1)
+    log_onehot = torch.log(x_onehot.float().clamp(min=1e-30))
+    return log_onehot
+
+
+@torch.jit.script
+def log_sub_exp(a: Tensor, b: Tensor) -> Tensor:
+    m = torch.maximum(a, b)
+    return torch.log(torch.exp(a - m) - torch.exp(b - m)) + m
+
+
+@torch.jit.script
+def sliced_logsumexp(x: Tensor, slices: Tensor) -> Tensor:
+    lse = torch.logcumsumexp(
+        torch.nn.functional.pad(x, [1, 0, 0, 0], value=-float("inf")), dim=-1
+    )
+
+    slice_starts = slices[:-1]
+    slice_ends = slices[1:]
+
+    slice_lse = log_sub_exp(lse[:, slice_ends], lse[:, slice_starts])
+    slice_lse_repeated = torch.repeat_interleave(
+        slice_lse, slice_ends - slice_starts, dim=-1
+    )
+    return slice_lse_repeated
+
+
+class FoundNANsError(BaseException):
+    """Found NANs during sampling"""
+
+    def __init__(self, message: str = "Found NANs during sampling.") -> None:
+        super(FoundNANsError, self).__init__(message)
+
+
+class TensorDataLoader:
```

Comment:
Shouldn't this inherit ` torch.utils.data.DataLoader`?

Reply:
This has a different signature from the torch DataLoader. More specifically, its input is not a Dataset but the tensors instead. It is said by the original author to be much faster than the torch implementation. We can remove it and use the standard DataLoader instead.

Reply:
I have replaced it with the standard combination of TensorDataset and DataLoader.